phantazm-mob/src/main/java/com/github/phantazmnetwork/mob/config/MobModelConfigProcessor.java

```diff
@@ -0,0 +1,165 @@
+import com.github.steanky.ethylene.core.processor.ConfigProcessor;
+import net.kyori.adventure.key.Key;
+import net.kyori.adventure.text.Component;
+import net.minestom.server.entity.EquipmentSlot;
+import net.minestom.server.item.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.*;
+
+public class MobModelConfigProcessor implements ConfigProcessor<MobModel> {
+
+    private static final ConfigProcessor<EquipmentSlot> EQUIPMENT_SLOT_PROCESSOR = ConfigProcessor.enumProcessor(EquipmentSlot.class);
+
+    private static final ConfigProcessor<Key> KEY_PROCESSOR = AdventureConfigProcessors.key();
+
+    private final ConfigProcessor<MinestomDescriptor> descriptorProcessor;
+
+    private final ConfigProcessor<Goal> goalProcessor;
+
+    private final ConfigProcessor<Skill> skillProcessor;
+
+    private final ConfigProcessor<Component> componentProcessor;
+
+    private final ConfigProcessor<ItemStack> itemStackProcessor;
+
+    public MobModelConfigProcessor(@NotNull ConfigProcessor<MinestomDescriptor> descriptorProcessor,
+                                   @NotNull ConfigProcessor<Goal> goalProcessor,
+                                   @NotNull ConfigProcessor<Skill> skillProcessor,
+                                   @NotNull ConfigProcessor<Component> componentProcessor,
+                                   @NotNull ConfigProcessor<ItemStack> itemStackProcessor) {
+        this.descriptorProcessor = Objects.requireNonNull(descriptorProcessor, "descriptorProcessor");
+        this.goalProcessor = Objects.requireNonNull(goalProcessor,
+                "goalCreatorConfigProcessor");
+        this.skillProcessor = Objects.requireNonNull(skillProcessor, "skillProcessor");
+        this.componentProcessor = Objects.requireNonNull(componentProcessor, "componentProcessor");
+        this.itemStackProcessor = Objects.requireNonNull(itemStackProcessor,
+                "itemStackConfigProcessor");
+    }
+
+    @Override
+    public @NotNull MobModel dataFromElement(@NotNull ConfigElement element) throws ConfigProcessException {
+        Key key = KEY_PROCESSOR.dataFromElement(element.getElementOrThrow("key"));
+
+        MinestomDescriptor descriptor = descriptorProcessor.dataFromElement(element.getElement("descriptor"));
+
+        ConfigList goalGroupsList = element.getListOrThrow("goalGroups");
+        Collection<Collection<Goal>> goalGroups = new ArrayList<>(goalGroupsList.size());
+        for (ConfigElement goalGroupElement : goalGroupsList) {
+            if (!goalGroupElement.isList()) {
+                throw new ConfigProcessException("goal groups are not a list");
+            }
+
+            ConfigList goalsList = goalGroupElement.asList();
+            Collection<Goal> goals = new ArrayList<>(goalsList.size());
+            for (ConfigElement goalElement : goalsList) {
+                goals.add(goalProcessor.dataFromElement(goalElement));
+            }
+
+            goalGroups.add(List.copyOf(goals));
```

Comment:
why List.copyOf?

Reply:
prevent modification? (I mean that's the obvious but lol)